src/Backend/SgateShopgatePlugin/Components/User.php

```diff
@@ -0,0 +1,396 @@
+            session_id($sessionId);
+        }
+
+        $basket = $this->basket->sGetBasket();
+
+        $httpResponse->setHeader('Content-Type', 'application/json');
+
+        if (isset($hash)) {
+            $email = strtolower($this->Request()->getPost('email'));
+            $user = $this->verifyUser($email, $hash);
+            if (!empty($user['sErrorMessages'])) {
+                $httpResponse->setHttpResponseCode(401);
+                $httpResponse->setBody(json_encode($user));
+                $httpResponse->sendResponse();
+                exit();
+            } else {
+                $httpResponse->setHttpResponseCode(200);
+                $httpResponse->setBody(json_encode(array(
+                    'id' => $user['customernumber'],
+                    'mail' => $user['email'],
+                    'first_name' => $user['firstname'],
+                    'last_name' => $user['lastname'],
+                    'birthday' => $user['birthday'],
+                    'customer_groups' => $user['customergroup'],
+                    'session_id' => $user['sessionID']
+                )));
+                $httpResponse->sendResponse();
+                exit();
+            }
+        } else {
+            $error = $this->admin->sLogin();
+        }
+
+        if (!empty($error['sErrorMessages'])) {
+            $httpResponse->setHttpResponseCode(401);
+            $httpResponse->setBody(json_encode($error));
+        } else {
+            if (!empty($basket['content'])) {
+                $this->basket->clearBasket();
+                $this->basket->sRefreshBasket();
+
+                foreach ($basket['content'] as $basketItem) {
+                    $this->basket->sAddArticle($basketItem['ordernumber'], $basketItem['quantity']);
+                }
+            }
+
+            $user = $this->admin->sGetUserData();
+            $user = $user['additional']['user'];
+
+            $httpResponse->setHttpResponseCode(200);
+            $httpResponse->setBody(json_encode(array(
+                'id' => $user['customernumber'],
+                'mail' => $user['email'],
+                'first_name' => $user['firstname'],
+                'last_name' => $user['lastname'],
+                'birthday' => $user['birthday'],
+                'customer_groups' => $user['customergroup'],
+                'session_id' => $user['sessionID']
+            )));
+        }
+
+        $this->basket->sRefreshBasket();
+
+        $httpResponse->sendResponse();
+        exit();
+    }
+
+    /**
+     * Custom get user action
+     *
+     * @param Enlight_Controller_Request_Request $request
+     */
+    public function getUser($request)
+    {
+        try {
+            $decoded = $this->webCheckoutHelper->getJWT($request->getCookie('token'));
+            $customerId = $decoded['customer_id'];
+
+            $sql = ' SELECT password, email FROM s_user WHERE customernumber = ? AND active=1 AND (lockeduntil < now() OR lockeduntil IS NULL) ';
```

Comment:
this can be sped up significantly if you give customernumber an index (it's varchar(30) so index size should be ok), you would need to write a migration for it in your Bootstrap.php. Runs in 300-500ms on a large s_user table and sub 1ms with an index

Reply:
Thank you for your input.
This table is created within the Shopware installation and maintained during updates. As we can't rule out side effects, if we alter that table, we don't want to do it